include/lp_hash_table.h

```diff
@@ -0,0 +1,100 @@
+#ifndef DIL_LP_HASH_TABLE_H
+#define DIL_LP_HASH_TABLE_H
+
+#include <stdlib.h>
+#include <stdint.h>
+#include <assert.h>
+
+template <typename T>
+struct lp_hash_table {
+private:
+    const char **keys;
+    T *values;
+    int buckets;
+
+    unsigned int hash(const char *key) {
+        // TODO: implement a better hashing method
+        return ((uintptr_t)key % buckets);
+    }
+
+public:
+
+    /*
+     * Constructor. Assumes that `elems > 0`
+     */
+    lp_hash_table(int elems) {
+        assert(elems > 0);
+
+        // TODO: maybe find a way to optimize this
+        buckets = elems;
+
+        // TODO: use our own allocator when it is ready.
+        keys = (const char **)malloc(buckets * sizeof(const char *));
+        values = (T *)malloc(buckets * sizeof(T));
+
+        for (int i = 0; i <= buckets; ++i) {
+            // We need to initialize the pointer array because the insertion loop will check on it's indices
+            keys[i] = nullptr;
+        }
+    };
+
+    /*
+     * @return - 0 means everything went ok, -1 means there was an error
+     */
+    int insert(const char *key, T value) {
+
+        int index = this->hash(key);
+        int i = index;
+
+        while(this->keys[i] != nullptr) {
+            if (this->keys[i] == key) {
+                return -1;
+            }
+
+            i++;
+            if (i == this->buckets)
+                i = 0; // Wrap around at the start
+            if (i == index)
+                return -1; // We have traversed the whole table and found no spot to insert in
+        }
+
+        this->keys[i] = key;
+        this->values[i] = value;
+        return 0;
+    };
+
+    /*
+     * @return - 0 everything went ok. -1 something went wrong
+     * @param value - the value of the element under `key` or nullptr if it does not exist
+     */
+    int find(const char *key, T *value) {
```

Comment:
`find` can just return the element. Since we constrain this hash table to only keep pointers, if didn't find it, we can just reuturn `NULL`. Otherwise, the found pointer. So, `T find(const char *key) { ... }`.

Reply:
I thought about this, actually it was the original implementation, but I liked the consistency the API would have in it's signatures. Especially if we used enum error codes and stuff like `perror` like the standard library. Moreover I didn't constrain `T` to be a pointer (bear in mind that I haven't used C++ in 2 years) so I guess by doing that returning the pointer here would be a sane choice.

Reply:
> Moreover I didn't constrain T to be a pointer

Of course. This wasn't the case as you initially did it, so returning NULL was not an option.

> Especially if we used enum error codes and stuff like perror like the standard library

Oh no, please! :) The `perror` et al of error handling is really bad. They use a global variable for all that and it's just a pain to work with. For `enum` codes, that's interesting, but here basically there are 3 cases we care about:
1) The value was found, return it.
2) The value was not found, return `NULL`.
3) Something went wrong, `assert`, halt and catch fire. :) (because something should never go wrong).

Reply:
Fair enough. I'll do it that way